src/utils/BucketUtils.ts

```diff
@@ -0,0 +1,384 @@
+    return {
+      result: 'Unable to determine a system match',
+      createDate: undefined,
+      rationales: ['Algorithm matched both true and false for this combination of datasets'],
+      id: bucket.key,
+    }
+  }
+}
+
+/**
+ * Process the match results for V3 Abstain. If we have a V3 result and we have
+ * an ABSTAIN case, we can return true if the number of abstentions > 0
+ * @param matchResults
+ */
+const processV3Abstain = (matchResults: MatchResult[]): boolean => {
+  const abstainList = map((m: MatchResult) => m.abstain)(matchResults)
+  const abstainValList = filter((a: boolean | undefined) => a === true)(abstainList)
+  return abstainValList.length > 0
+}
+
+/**
+ * Calculate "Other" status for a data use. Data Uses can have 'otherRestrictions': TRUE|FALSE,
+ * or they can have fields populated for 'other': 'other restriction' and 'secondaryOther': 'yet other restriction'
+ * @private
+ * @param dataUse
+ * @returns boolean
+ */
+const isOther = (dataUse?: DataUseSummary): boolean => {
+  const primaryOther = dataUse?.primary?.some((dut: DataUseTerm) => dut.code === 'OTHER') || false
+  const secondaryOther = dataUse?.secondary?.some((dut: DataUseTerm) => dut.code === 'OTHER') || false
+  return primaryOther || secondaryOther
+}
+
+/**
+ * Calculate abstention for a data use. There are a number of cases where there should
+ * not be an algorithm decision if a field is true, including any "Other" state.
+ * @param dataUse
+ * @returns boolean
+ */
+export const shouldAbstain = (dataUse?: DataUseSummary): boolean => {
+  const codeList: string[] = ['OTHER', 'POP-M', 'POP-F', 'COL', 'IRB', 'GSO', 'PUB', 'MOR', 'POP-PD']
+  return isOther(dataUse) || dataUse?.secondary?.some((dut: DataUseTerm) => {
+    return codeList.some((code: string) => code === dut.code)
+  }) || false
+}
+
+/**
+ * Create a structure of RP (research purpose) votes from all votes in a list of buckets.
+ *
+ * @private
+ * @param buckets
+ * @returns {Array<{rp: VoteGroup}>}
+ */
+const createRpVoteStructureFromBuckets = (buckets: Bucket[]): Array<{ rp: VoteGroup }> => {
+  // List of rp vote groups broken out by election into chair, member, and final votes.
+  const rpVotes: Array<{ rp: VoteGroup }> = []
+
+  const rpElectionVoteArrays: Vote[][] = flow(
+    flatMap((b: Bucket) => b.elections),
+    filter((e: Election) => e.electionType.toLowerCase() === 'rp'),
+    map((e: Election) => e.votes),
+    // election.votes is a hash of vote id => vote object
+    map((hash: Record<string, Vote>) => values(hash)),
+  )(buckets)
+
+  forEach((vArray: Vote[]) => {
+    const rpVoteGroup: VoteGroup = {
+      chairpersonVotes: [],
+      memberVotes: [],
+      finalVotes: [],
+    }
+    forEach((v: Vote) => {
+      const lowerCaseType = v.type.toLowerCase()
+      switch (lowerCaseType) {
+        case 'chairperson':
+          // 'Chairperson' votes count as final votes for 'RP' elections. This is not true for 'DataAccess' votes
```

Comment:
Helpful clarification!  I mistakenly thought Chair votes were automatically final in _all_ types of elections.  

Is this more correct?

```suggestion
          // 'Chairperson' votes count as final votes for 'RP' elections. This is not true for 'DataAccess' elections
```

Reply:
I had to go back to the back end code to verify this, but yes, Chair votes do count as final for RP elections. We don't capture a FINAL vote type for RP elections, we only do that for DataAccess elections. Good language clarification, thank you 👍🏽